ask: Fix bug with algebraic exponential handler

krishnavbajoria02 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #27445


#### Brief description of what is fixed or changed
This pull request fixes issue #27445, where ask(Q.algebraic(Pow(1, e, evaluate=False))) incorrectly returns False. The problem arises because the current implementation of the Q.algebraic handler for Pow expressions does not account for the special case where the base is 1.

#### Other comments
This fix ensures that the mathematical properties of \( 1^b \) are respected, which always evaluates to 1, and therefore should be considered algebraic.


#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* assumptions
  * Fixed a bug in the `ask` function where `Q.algebraic` queries did not handle three-valued logic correctly. This ensures the behavior is consistent when handling cases with undefined truth values.
<!-- END RELEASE NOTES -->

krishnavbajoria02 committed:
Fixed algebraic power handler bug, added additional checks and regression tests

- Fixed the incorrect three-value logic in the algebraic power handler.
- Added additional conditions based on the Gelfond-Schneider theorem and other special cases.
- Implemented extensive regression tests to cover various edge cases and ensure robustness.
- Optimized the efficiency of the power handling process to improve performance.

TiloRC commented:
You should add a release notes entry.
https://github.com/sympy/sympy/wiki/Writing-Release-Notes

TiloRC commented:
See Oscar's comment for the right fix: https://github.com/sympy/sympy/issues/27445#issuecomment-2614147352.

You should add Oscar as a coauthor. See https://docs.sympy.org/latest/contributing/new-contributors-guide/workflow-process.html#co-author.

oscarbenjamin commented:
> You should add Oscar as a coauthor.

It isn't necessary to add me as a co-author. I don't need any acknowledgement.

TiloRC commented:
Oh, you should also add a regression test. See https://docs.sympy.org/latest/contributing/new-contributors-guide/workflow-process.html#workflow-process-add-tests.

krishnavbajoria02 commented:
I have added the regression test and fixed the code issue. Please review it.

TiloRC commented on sympy/assumptions/tests/test_query.py:
```diff
@@ -2021,6 +2021,9 @@ def test_algebraic():
     assert ask(Q.algebraic(x), Q.transcendental(x)) is False
     assert ask(Q.transcendental(x), Q.algebraic(x)) is False
 
+    e = Symbol('e', algebraic=True)
```
This is my fault, but I don't think e is a good variable name as it can be confused with Euler's number. Just using "x" is better. Also, it matches the style of the file better if you avoid using the old assumptions. So something like the following might be a better test:
```python
ask(Q.algebraic(Pow(1, x, evaluate=False)), Q.algebraic(x)) is True
```

Also, I would add the following test:
```python
assert ask(Q.algebraic(Pow(x, y))) is None
```
This also wrongly gives false.

And also add:
```python
ask(Q.algebraic(Pow(1, x, evaluate=False))) is None
```
If x is an imaginary number, the above expression is undefined.

TiloRC commented on sympy/assumptions/tests/test_query.py:
```diff
@@ -2021,6 +2021,10 @@ def test_algebraic():
     assert ask(Q.algebraic(x), Q.transcendental(x)) is False
     assert ask(Q.transcendental(x), Q.algebraic(x)) is False
 
```
You should add a comment with a link to the GitHub issue as described here: https://docs.sympy.org/latest/contributing/new-contributors-guide/writing-tests.html#writing-tests-regression-tests

krishnavbajoria02 replied:
@TiloRC 
I have added the comment as shown in the link.

TiloRC commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -737,11 +737,17 @@ def _(expr, assumptions):
-    return expr.exp.is_Rational and ask(Q.algebraic(expr.base), assumptions)
+    if expr.exp.is_Rational:
+        return ask(Q.algebraic(expr.base), assumptions)
+    else:
+        return False
```
This is incorrect and should be deleted. For example, $e^{2\pi i} = 1$ has a non-rational exponent but is an algebraic number.  Can you also add the following test?
```python
assert ask(Q.algebraic(x**(pi*I))) is None
```

krishnavbajoria02 replied:
@TiloRC
 I have made the changes and added the test.

TiloRC commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -737,11 +737,17 @@ def _(expr, assumptions):
 @AlgebraicPredicate.register(Pow) # type:ignore
 def _(expr, assumptions):
+    if expr.base == 1:
+        return True
```
This isn't correct. For example, $1^\infty$ is not Algebraic. Variables in SymPy aren't assumed to be finite by default.

krishnavbajoria02 replied:
@TiloRC I have made the changes. Please review it once.

TiloRC commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -737,11 +737,23 @@ def _(expr, assumptions):
+        if expr.exp.as_real_imag()[1]!=0:
+            return None
+        return False
+    else:
+        return None
```
This is redundant. If a function doesn't explicitly return something, None is returned by default.

TiloRC commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -737,11 +737,23 @@ def _(expr, assumptions):
 
 @AlgebraicPredicate.register(Pow) # type:ignore
 def _(expr, assumptions):
+    if expr.base == 1:
+        if ask(Q.infinite(expr.exp), assumptions)==False:
+            return True
+        return None
```
I think the added code complexity is not worth the tiny benefit of being able to handle this obscure case.

krishnavbajoria02 replied:
@TiloRC 
I was trying to make it a simpler condition but I couldn't do better than this. Can you guide me how to move ahead here or should I delete this condition and return None so as to not add further complexity to the code? 
Thanks.

TiloRC replied:
These particular lines of code can be deleted. We don't want to remove any existing functionality though.

krishnavbajoria02 replied:
@TiloRC Since we are not considering the edge case of 1^infinity, should I simply return `True` here? The original issue arose because the handler did not account for the fact that any power of 1 remains 1, meaning `False` should not be returned.

TiloRC replied:
Don't check if the base is 1 or not. Delete the whole thing.

krishnavbajoria02 replied:
@TiloRC
```assert ask(Q.algebraic(Pow(1, x, evaluate=False)), Q.algebraic(x)) is True```
I will do that but then for this test case either we have to change True to None or remove the test altogether as to return True then we require both the base to be algebraic and the exponent to be rational.

krishnavbajoria02 commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -737,11 +737,23 @@ def _(expr, assumptions):
-    return expr.exp.is_Rational and ask(Q.algebraic(expr.base), assumptions)
+    if ask(Q.rational(expr.exp),assumptions) and ask(Q.algebraic(expr.base),assumptions):
+        return True
+    elif ask(Q.rational(expr.exp),assumptions)==False or ask(Q.algebraic(expr.base),assumptions)==False:
```
@TiloRC 
I have made the changes. Please review it.

TiloRC commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -737,11 +737,22 @@ def _(expr, assumptions):
+    if exp_rational and base_algebraic:
+        return True
+    elif exp_rational is False or base_algebraic is False:
+        if expr.exp.as_real_imag()[1]!=0:
```
what's the logic behind this? You're checking if the exponenet is an imaginary number? It's better not to check for spesific values like this. We want to be able to handle symbolic variables, and if we can handle those we'll be able to handle constants. However, if we just handle constants we might get wrong results for symbolic variables.

krishnavbajoria02 replied:
@TiloRC 
I have made the changes so that it works for symbolic variables also. Please do review it.

TiloRC commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -737,11 +737,22 @@ def _(expr, assumptions):
+    base_algebraic = ask(Q.algebraic(expr.base), assumptions)
+    if exp_rational and base_algebraic:
+        return True
+    elif exp_rational is False or base_algebraic is False:
+        if ask(Q.imaginary(expr.exp), assumptions):
+            return None
+        return False
```
I expect the following tests to fail because `ask` returns False instead of None. 
```python
# x**y could be undefined if y was not finite
assert ask(Q.algebriac(x**y), ~Q.algebraic(x) & ~Q.rational(y)) is None
# x**y could be algebriac if y=0
assert ask(Q.algebriac(x**y), Q.transcendental(x) & Q.real(y)) is None
# x**y could be algebriac if x=0
assert ask(Q.algebriac(x**y), & Q.real(x) & Q.irrational(y)) is None
```
They would make good additions to the tests.

TiloRC replied:
Also, we forgot about another edge case:
```python
>>> 2**log(3, 2)
3
```

TiloRC replied:
https://en.wikipedia.org/wiki/Gelfond%E2%80%93Schneider_theorem

krishnavbajoria02 replied:
@TiloRC I have added the check for certain 'False' cases. Please do review it.

krishnavbajoria02 commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -741,7 +741,14 @@ def _(expr, assumptions):
+    if base_algebraic and exp_algebraic:
+      if exp_rational:
+          return True
+      elif ask(Q.ne(expr.base,0)) and ask(Q.ne(expr.base,1)) and exp_rational is False:
```
@TiloRC 
I have added the comment as well as made the code changes as you have told to do. Please review them.

krishnavbajoria02 replied:
@TiloRC 
I am sorry for this.. I am relatively new to open source so have been facing this issue. I was trying to run locally only by trying to follow these instructions but I think every time I make changes to this branch the test automatically gets called here..do I have to create another branch to test it (using pytest) as I am not sure about this.

TiloRC commented:
pi**n is known to be algebraic for positive integers:
- https://math.stackexchange.com/questions/1875840/references-on-pin-is-transcendental?noredirect=1&lq=1
- https://math.stackexchange.com/questions/31798/prove-that-pi-is-a-transcendental-number

We can add a check for this like the existing check for e

krishnavbajoria02 commented:
> pi**n is known to be algebraic for positive integers:

@TiloRC 
You mean not algebraic right?

TiloRC commented:
I would recommend running the tests locally on your computer: https://docs.sympy.org/latest/contributing/new-contributors-guide/workflow-process.html#workflow-process-run-tests

krishnavbajoria02 commented:
@TiloRC 
I am sorry for this.. I am relatively new to open source so have been facing this issue. I was trying to run locally only by trying to follow these instructions but I think every time I make changes to this branch the test automatically gets called here..do I have to create another branch to test it (using pytest) as I am not sure about this.

krishnavbajoria02 commented on sympy/assumptions/tests/test_query.py:
```diff
@@ -2013,14 +2013,21 @@ def test_algebraic():
         assert ask(Q.algebraic(h(x)), Q.algebraic(x)) is False
 
-    assert ask(Q.algebraic(sqrt(sin(7)))) is False
+    assert ask(Q.algebraic(sqrt(sin(7)))) is None
```
@TiloRC 
Since it is sqrt() so it will do something like sqrt(x) = x**1/2 which will require the Pow handler.

krishnavbajoria02 replied:
@TiloRC 
for more reference this can be useful:

```
>> from sympy import sqrt, symbols, Pow, S
>> x = symbols('x')
>> expr1 = sqrt(x)
>> expr2 = Pow(x, S.Half) 
>> print(expr1 == expr2) 
>> True
```

TiloRC replied:
> >> print(expr1 == expr2) 

Just because they're equal, doesn't mean they're represented the same way or would use the same handlers.

TiloRC replied:
But you are right:
```python
>>> from sympy import sqrt, symbols, Pow, S
>>> x = symbols('x')
>>> expr1 = sqrt(x)
>>> type(expr1)
<class 'sympy.core.power.Pow'>
```

krishnavbajoria02 replied:
@TiloRC
https://github.com/sympy/sympy/blob/4c7eafd523abea2fa402e2e6da54beaa69921ca1/sympy/codegen/cfunctions.py#L385-L386

TiloRC replied:
I think we can still handle these sorts of cases.

https://math.stackexchange.com/questions/112938/when-is-sin-x-an-algebraic-number-and-when-is-it-non-algebraic

https://math.stackexchange.com/questions/5030029/is-1-pin-for-all-integer-n-1-known-to-be-transcendental-what-about-x

TiloRC replied:
It might make sense to move the logic that handles constant numbers to another function like this:

https://github.com/sympy/sympy/blob/4c7eafd523abea2fa402e2e6da54beaa69921ca1/sympy/assumptions/handlers/sets.py#L30-L38

krishnavbajoria02 replied:
@TiloRC
I think for the sin case then adding this condition would fix it (but again it is handling a specific case only so I am not sure if this is the right way to go about or not).
```
if isinstance(expr.base, sympy.Function) and expr.base.func == sympy.sin:
        if ask(Q.algebraic(expr.base.args[0]), assumptions) and ask(Q.ne(expr.base.args[0],0),assumptions):  
            return False
```
Now regarding the ```(1+pi)**2``` case it is kind of suggesting if the power is a positive integer and the base is transcendental(since polynomial is just a mix of additions and multiplications so asking something about the base will call for the Add and Mul handlers here) then we can conclude that the expression is transcendental. If that is the case I think this proposal should work?
```
if ask(Q.integer(expr.pow),assumptions) and ask(Q.positive(expr.pow),assumptions): 
  if ask(Q.algebraic(expr.base),assumptions) is False:
     return False

TiloRC commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -741,7 +741,20 @@ def _(expr, assumptions):
+          return True
+      # Check based on the Gelfond-Schneider theorem:
+      # If the base is algebraic and not equal to 0 or 1, and the exponent is irrational,then the result is transcendental.
+      elif ask(Q.ne(expr.base,0) & Q.ne(expr.base,1)) and exp_rational is False:
```
this doesn't need to be an elif statement; it can just be an if statement

krishnavbajoria02 replied:
@TiloRC 
Right right.. I have made the changes. Please do review them.

TiloRC commented on sympy/assumptions/handlers/sets.py:
```diff
@@ -741,7 +741,20 @@ def _(expr, assumptions):
+        return
+    exp_rational = ask(Q.rational(expr.exp), assumptions)
+    base_algebraic = ask(Q.algebraic(expr.base), assumptions)
+    exp_algebraic = ask(Q.algebraic(expr.exp),assumptions)
+    if base_algebraic and exp_algebraic:
+      if exp_rational:
+          return True
+      # Check based on the Gelfond-Schneider theorem:
+      # If the base is algebraic and not equal to 0 or 1, and the exponent is irrational,then the result is transcendental.
```
Nice comment!

oscarbenjamin replied:
Please use 4 spaces for indentation as in PEP 8.

Also there is no need for comments to have such long lines. They can be wrapped at 80 characters.

krishnavbajoria02 commented on sympy/assumptions/tests/test_query.py:
```diff
@@ -420,7 +420,7 @@ def test_pi():
     assert ask(Q.integer(z)) is False
     assert ask(Q.rational(z)) is False
-    assert ask(Q.algebraic(z)) is False
+    assert ask(Q.algebraic(z)) is None
```
@TiloRC I think this is happening because of the following code not breaking the 'whole-square' into individual parts 
( (1+pi)**2 = 1 + 2*pi + pi^2) for which reason this is not being given the False value.
```
>> from sympy import pi
>>> expr=(1 + pi) ** 2
>>> expr.simplify()
(1 + pi)**2
```

krishnavbajoria02 commented:
@TiloRC 
Any more changes required ?

TiloRC commented on sympy/assumptions/tests/test_query.py:
```diff
@@ -2013,14 +2013,21 @@ def test_algebraic():
         assert ask(Q.algebraic(h(x)), Q.algebraic(x)) is False
 
-    assert ask(Q.algebraic(sqrt(sin(7)))) is False
+    assert ask(Q.algebraic(sqrt(sin(7)))) is None
```
I think for now this is fine. I'll open up an issue about handling these sorts of cases.

TiloRC replied:
@oscarbenjamin 

Ideally, `ask(Q.algebraic(sqrt(sin(7))))` would return False. This PR makes it return None instead.

I believe a consequence of the [Lindemann–Weierstrass theorem](http://en.wikipedia.org/wiki/Lindemann%E2%80%93Weierstrass_theorem) is that for all non-constant polynomial $f$, the following are transcendetal:
- $f(\pi)$
- $f(e)$
- $f(sin(c))$ where $sin(c)$ is transcendetal (and the same for other trig functions)

Should we implement that behavior here in this PR, or allow for the test `ask(Q.algebraic(sqrt(sin(7))))` to fail and open up a new issue to implement this functionality?

oscarbenjamin commented:
Please use commit messages that explain what is actually changed. None of these commit messages explains what is changed:
https://github.com/sympy/sympy/pull/27511/commits

krishnavbajoria02 commented:
@oscarbenjamin @TiloRC 
I have updated the commit messages. Please review them.

oscarbenjamin commented:
It would be better at this point to squash all the commits into one commit with a single commit message.

krishnavbajoria02 commented:
@oscarbenjamin I have squashed all commits into one. Please review them.

TiloRC commented:
@krishnavbajoria02 see https://www.freecodecamp.org/news/git-squash-commits/


Basically, run `git rebase -i HEAD~32` (32 since there are 32 commits right now) in the terminal. You'll get something that looks like this:

<img width="587" alt="Screenshot 2025-02-02 at 3 59 08 PM" src="https://github.com/user-attachments/assets/f553b994-e22d-48c6-a3ea-92d75e55eaef" />

Then edit the file to move the mailmap commits to the top and "2ee8a0c5b0 fixes #27445 by accounting for base 1" underneath them. Then for all of the commits under "2ee8a0c5b0 fixes #27445 by accounting for base 1", delete "pick" and replace it with "s" or "squash". 

It might be a bit tricky to edit, because the editor that it opens up by default is vi. See [this](https://www.redhat.com/en/blog/introduction-vi-editor) guide on vi for how to use it. But the basic commands are:

i — Switch to Insert mode.
Esc — Switch to Command mode.
:wq or ZZ — Save and quit/exit vi.

In insert mode you'll be able to edit the file, and in command mode you'll be able to use :wq to quit.

krishnavbajoria02 commented:
@TiloRC 
I think I have done that only. Can you please review it once. Thanks.

TiloRC commented on sympy/assumptions/tests/test_query.py:
```diff
@@ -2013,14 +2013,21 @@ def test_algebraic():
     assert ask(Q.algebraic(x), Q.transcendental(x)) is False
     assert ask(Q.transcendental(x), Q.algebraic(x)) is False
 
+    #https://github.com/sympy/sympy/issues/27445
+    assert ask(Q.algebraic(Pow(1, x, evaluate=False)), Q.algebraic(x)) is None
+    assert ask(Q.algebraic(Pow(x, y))) is None
+    assert ask(Q.algebraic(Pow(1, x, evaluate=False))) is None
+    assert ask(Q.algebraic(x**(pi*I))) is None
+    assert ask(Q.algebraic(pi**n),Q.integer(n) & Q.positive(n)) is False
+
```
Can you add a test where the exponent is rational that asserts True so we can get full code coverage?

TiloRC replied:
After that I think it should be ready to merge

krishnavbajoria02 replied:
@TiloRC 
I have added that test. Please review it.

krishnavbajoria02 committed:
Added regression test in test_query.py

Added the test for x**y when x is algebraic and y is rational which should return true to get entire code coverage.

krishnavbajoria02 commented:
@oscarbenjamin @TiloRC all tests are passing now. Can you please review it.

TiloRC commented:
Good work!

krishnavbajoria02 commented:
@TiloRC @oscarbenjamin 
I would like to sincerely thank you both for taking the time to guide me through my first contribution. I truly appreciate your effort in reviewing my work and helping me improve. Looking forward to contributing more and learning from this amazing community!
